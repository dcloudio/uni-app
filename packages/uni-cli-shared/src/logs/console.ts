import MagicString from 'magic-string'
import type { TransformResult } from 'vite'
import { normalizePath } from '../utils'

export function rewriteConsoleExpr(
  method: string,
  id: string,
  filename: string,
  code: string,
  sourceMap: boolean = false
): TransformResult {
  filename = normalizePath(filename)
  const re = /(console\.(log|info|debug|warn|error))\s*\(([^)]+)\)/g
  const locate = getLocator(code)
  const s = new MagicString(code)
  let match: RegExpExecArray | null
  while ((match = re.exec(code))) {
    const [, expr, type] = match
    s.overwrite(
      match.index,
      match.index + expr.length + 1,
      method + `('${type}','at ${filename}:${locate(match.index).line + 1}',`
    )
  }
  if (s.hasChanged()) {
    return {
      code: s.toString(),
      map: sourceMap ? s.generateMap({ hires: true }) : { mappings: '' },
    }
  }
  return { code, map: { mappings: '' } }
}

export function restoreConsoleExpr(code: string): string {
  return code.replace(
    /(?:uni\.)?__f__\('([^']+)','at ([^:]+):(\d+)',/g,
    'console.$1('
  )
}

function getLocator(source: string) {
  const originalLines = source.split('\n')
  const lineOffsets: number[] = []

  for (let i = 0, pos = 0; i < originalLines.length; i++) {
    lineOffsets.push(pos)
    pos += originalLines[i].length + 1
  }

  return function locate(index: number) {
    let i = 0
    let j = lineOffsets.length
    while (i < j) {
      const m = (i + j) >> 1
      if (index < lineOffsets[m]) {
        j = m
      } else {
        i = m + 1
      }
    }
    const line = i - 1
    const column = index - lineOffsets[line]
    return { line, column }
  }
}
