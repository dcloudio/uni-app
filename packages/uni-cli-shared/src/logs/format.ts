import { once } from '@dcloudio/uni-shared'
import { isString } from '@vue/shared'

import { isInHBuilderX, runByHBuilderX } from '../hbx/env'
import { moduleAliasFormatter } from '../hbx/alias'
import {
  errorFormatter,
  h5ServeFormatter,
  removeDuplicatePluginFormatter,
  removeInfoFormatter,
  removeWarnFormatter,
} from '../hbx/log'
import type { LogErrorOptions, LogOptions } from 'vite'

export interface Formatter<T extends LogOptions = LogOptions> {
  test: (msg: string, options?: T) => boolean
  format: (msg: string, options?: T) => string
}

const errFormatters: Formatter<LogErrorOptions>[] = []
const infoFormatters: Formatter[] = []
const warnFormatters: Formatter[] = []

const initErrFormattersOnce = once(() => {
  if (isInHBuilderX()) {
    errFormatters.push(moduleAliasFormatter)
  }
  errFormatters.push(errorFormatter)
})

const initInfoFormattersOnce = once(() => {
  if (runByHBuilderX()) {
    if (
      // 开发模式下
      process.env.UNI_PLATFORM === 'h5' &&
      process.env.NODE_ENV !== 'production'
    ) {
      infoFormatters.push(h5ServeFormatter)
    }
  }
  infoFormatters.push(removeInfoFormatter)
})

const initWarnFormattersOnce = once(() => {
  warnFormatters.push(removeWarnFormatter)
})

export function formatErrMsg(msg: string, options?: LogErrorOptions) {
  if (options && isInHBuilderX()) {
    options.timestamp = false
  }
  initErrFormattersOnce()
  const formatter = errFormatters.find(({ test }) => test(msg, options))
  if (formatter) {
    return formatter.format(msg, options)
  }
  return msg
}

const REMOVED_NVUE_MSGS = [
  (msg: string) => {
    // vite v2.7.10 building for development... (x2)
    return msg.includes('vite v') && msg.includes('building ')
  },
]

export const removeNVueInfoFormatter: Formatter = {
  test(msg) {
    return !!REMOVED_NVUE_MSGS.find((m) =>
      isString(m) ? msg.includes(m) : m(msg)
    )
  },
  format() {
    return ''
  },
}
const nvueInfoFormatters: Formatter[] = []
const initNVueInfoFormattersOnce = once(() => {
  nvueInfoFormatters.push(removeNVueInfoFormatter)
})

export function formatInfoMsg(
  msg: string,
  options?: LogOptions & { nvue?: boolean }
) {
  if (options && isInHBuilderX()) {
    options.timestamp = false
  }
  initInfoFormattersOnce()
  const formatter = infoFormatters.find(({ test }) => test(msg, options))
  if (formatter) {
    return formatter.format(msg, options)
  }
  if (options?.nvue) {
    initNVueInfoFormattersOnce()
    const formatter = nvueInfoFormatters.find(({ test }) => test(msg, options))
    if (formatter) {
      return formatter.format(msg, options)
    }
  }
  return msg
}

export function formatWarnMsg(msg: string, options?: LogOptions) {
  if (options && isInHBuilderX()) {
    options.timestamp = false
  }
  initWarnFormattersOnce()
  msg = removeDuplicatePluginFormatter.format(msg)
  const formatter = warnFormatters.find(({ test }) => test(msg, options))
  if (formatter) {
    return formatter.format(msg, options)
  }
  return msg
}
