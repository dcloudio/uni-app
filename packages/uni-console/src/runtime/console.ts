import type { ComponentInternalInstance, ComponentPublicInstance } from 'vue'
import type { SendFn } from './utils'

const CONSOLE_TYPES = ['log', 'warn', 'error', 'info', 'debug'] as const

type MessageType = 'log' | 'warn' | 'error' | 'info' | 'debug'

let sendConsole: SendFn = null

type Message = {
  type: MessageType
  args: Array<any>
}

const messageQueue: Message[] = []

function sendConsoleMessages(messages: Message[]) {
  if (sendConsole == null) {
    messageQueue.push(...messages)
    return
  }
  sendConsole(
    JSON.stringify({
      type: 'console',
      data: messages,
    })
  )
}

export function setSendConsole(value: SendFn) {
  sendConsole = value
  if (value != null && messageQueue.length > 0) {
    const messages = messageQueue.slice()
    messageQueue.length = 0
    sendConsoleMessages(messages)
  }
}

export const originalConsole = /*@__PURE__*/ CONSOLE_TYPES.reduce(
  (methods, type) => {
    methods[type] = console[type].bind(console)
    return methods
  },
  {} as Record<MessageType, typeof console.log>
)

const atFileRegex = /^at\s+[\w/./-]+:\d+$/

export function rewriteConsole() {
  function wrapConsole(type: MessageType) {
    return function (...args: any[]) {
      const originalArgs = [...args]
      if (originalArgs.length) {
        const maybeAtFile = originalArgs[originalArgs.length - 1]
        // 移除最后的 at pages/index/index.uvue:6
        if (typeof maybeAtFile === 'string' && atFileRegex.test(maybeAtFile)) {
          originalArgs.pop()
        }
      }
      if (__UNI_CONSOLE_KEEP_ORIGINAL__) {
        originalConsole[type](...originalArgs)
      }
      sendConsoleMessages([formatMessage(type, args)])
    }
  }

  // 百度小程序不允许赋值，所以需要判断是否可写
  if (isConsoleWritable()) {
    CONSOLE_TYPES.forEach((type) => {
      console[type] = wrapConsole(type)
    })
    return function restoreConsole() {
      CONSOLE_TYPES.forEach((type) => {
        console[type] = originalConsole[type]
      })
    }
  } else {
    // @ts-expect-error
    const oldLog = uni.__f__
    if (oldLog) {
      // 重写 uni.__f__ 方法，这样的话，仅能打印开发者代码里的日志，其他没有被重写为__f__的日志将无法打印（比如uni-app框架、小程序框架等）
      // @ts-expect-error
      uni.__f__ = function (...args: any[]) {
        const [type, filename, ...rest] = args
        // 原始日志移除 filename
        oldLog(type, '', ...rest)
        sendConsoleMessages([formatMessage(type, [...rest, filename])])
      }
      return function restoreConsole() {
        // @ts-expect-error
        uni.__f__ = oldLog
      }
    }
  }
  return function restoreConsole() {}
}

function isConsoleWritable() {
  const value = console.log
  const sym = Symbol()
  try {
    // @ts-expect-error
    console.log = sym
  } catch (ex) {
    return false
  }
  // @ts-expect-error
  const isWritable = console.log === sym
  console.log = value
  return isWritable
}

function formatMessage(type: MessageType, args: Array<any | null>): Message {
  try {
    return {
      type,
      args: formatArgs(args),
    }
  } catch (e) {
    originalConsole.error(e)
  }
  return {
    type,
    args: [],
  }
}

function formatArgs(args: Array<any | null>) {
  return args.map((arg) => formatArg(arg))
}

export function formatArg(arg: any | null, depth: number = 0) {
  if (depth >= 7) {
    return {
      type: 'object',
      value: '[Maximum depth reached]',
    }
  }
  return ARG_FORMATTERS[typeof arg](arg, depth)
}

function formatObject(value: object, depth: number) {
  if (value === null) {
    return {
      type: 'null',
    }
  }

  if (isComponentPublicInstance(value)) {
    return formatComponentPublicInstance(value, depth)
  }

  if (isComponentInternalInstance(value)) {
    return formatComponentInternalInstance(value, depth)
  }

  if (isUniElement(value)) {
    return formatUniElement(value, depth)
  }

  if (isCSSStyleDeclaration(value)) {
    return formatCSSStyleDeclaration(value, depth)
  }

  if (Array.isArray(value)) {
    return {
      type: 'object',
      subType: 'array',
      value: {
        properties: value.map((v, i) => formatArrayElement(v, i, depth + 1)),
      },
    }
  }
  if (value instanceof Set) {
    return {
      type: 'object',
      subType: 'set',
      className: 'Set',
      description: `Set(${value.size})`,
      value: {
        entries: Array.from(value).map((v) => formatSetEntry(v, depth + 1)),
      },
    }
  }
  if (value instanceof Map) {
    return {
      type: 'object',
      subType: 'map',
      className: 'Map',
      description: `Map(${value.size})`,
      value: {
        entries: Array.from(value.entries()).map((v) =>
          formatMapEntry(v, depth + 1)
        ),
      },
    }
  }

  if (value instanceof Promise) {
    return {
      type: 'object',
      subType: 'promise',
      value: {
        properties: [],
      },
    }
  }
  if (value instanceof RegExp) {
    return {
      type: 'object',
      subType: 'regexp',
      value: String(value),
      className: 'Regexp',
    }
  }
  if (value instanceof Date) {
    return {
      type: 'object',
      subType: 'date',
      value: String(value),
      className: 'Date',
    }
  }
  if (value instanceof Error) {
    return {
      type: 'object',
      subType: 'error',
      value: value.message || String(value),
      className: value.name || 'Error',
    }
  }
  return {
    type: 'object',
    value: {
      properties: Object.entries(value).map(([name, value]) =>
        formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function isComponentPublicInstance(
  value: any
): value is ComponentPublicInstance {
  return value.$ && isComponentInternalInstance(value.$)
}

function isComponentInternalInstance(
  value: any
): value is ComponentInternalInstance {
  return value.type && value.uid != null && value.appContext
}

function formatComponentPublicInstance(
  value: ComponentPublicInstance,
  depth: number
) {
  return {
    type: 'object',
    className: 'ComponentPublicInstance',
    value: {
      properties: Object.entries(value.$.type).map(([name, value]) =>
        formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function formatComponentInternalInstance(
  value: ComponentInternalInstance,
  depth: number
) {
  return {
    type: 'object',
    className: 'ComponentInternalInstance',
    value: {
      properties: Object.entries(value.type).map(([name, value]) =>
        formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function isUniElement(value: any): value is UniElement {
  return value.style && value.tagName != null && value.nodeName != null
}

function formatUniElement(value: UniElement, depth: number) {
  return {
    type: 'object',
    // 非 x 没有 UniElement 的概念
    // className: 'UniElement',
    value: {
      properties: Object.entries(value)
        .filter(([name]) =>
          [
            'id',
            'tagName',
            'nodeName',
            'dataset',
            'offsetTop',
            'offsetLeft',
            'style',
          ].includes(name)
        )
        .map(([name, value]) => formatObjectProperty(name, value, depth + 1)),
    },
  }
}

function isCSSStyleDeclaration(
  value: any
): value is CSSStyleDeclaration & { $styles: Record<string, string | null> } {
  return (
    typeof value.getPropertyValue === 'function' &&
    typeof value.setProperty === 'function' &&
    value.$styles
  )
}

function formatCSSStyleDeclaration(
  style: CSSStyleDeclaration & { $styles: Record<string, string | null> },
  depth: number
) {
  return {
    type: 'object',
    value: {
      properties: Object.entries(style.$styles).map(([name, value]) =>
        formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function formatObjectProperty(name: string, value: any | null, depth: number) {
  return Object.assign(formatArg(value, depth), {
    name,
  })
}
function formatArrayElement(value: any | null, index: number, depth: number) {
  return Object.assign(formatArg(value, depth), {
    name: `${index}`,
  })
}

function formatSetEntry(value: any | null, depth: number) {
  return {
    value: formatArg(value, depth),
  }
}

function formatMapEntry(value: Array<any | null>, depth: number) {
  return {
    key: formatArg(value[0], depth),
    value: formatArg(value[1], depth),
  }
}

const ARG_FORMATTERS = {
  function(value: any) {
    return {
      type: 'function',
      value: `function ${value.name}() {}`,
    }
  },
  undefined() {
    return {
      type: 'undefined',
    }
  },
  object(value: object, depth: number) {
    return formatObject(value, depth)
  },
  boolean(value: boolean) {
    return {
      type: 'boolean',
      value: String(value),
    }
  },
  number(value: number) {
    return {
      type: 'number',
      value: String(value),
    }
  },
  bigint(value: bigint) {
    return {
      type: 'bigint',
      value: String(value),
    }
  },
  string(value: string) {
    return {
      type: 'string',
      value,
    }
  },
  symbol(value: symbol) {
    return {
      type: 'symbol',
      value: value.description,
    }
  },
}
