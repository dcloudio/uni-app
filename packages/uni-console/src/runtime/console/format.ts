import type { ComponentInternalInstance, ComponentPublicInstance } from 'vue'

interface NormalizeResult {
  name?: string
  type: string
  subType?: string
  className?: string
  description?: string
  value?: any
}

interface ObjectResultValue {
  properties: Array<NormalizeResult>
}
interface ObjectResult extends NormalizeResult {
  value: ObjectResultValue
}
interface ArrayResultValue {
  properties: Array<NormalizeResult>
}
interface ArrayResult extends NormalizeResult {
  value: ArrayResultValue
}
interface SetResultEntry {
  value: NormalizeResult
}
interface SetResultValue {
  entries: Array<SetResultEntry>
}
interface SetResult extends NormalizeResult {
  value: SetResultValue
}
interface MapResultEntry {
  key: NormalizeResult
  value: NormalizeResult
}
interface MapResultValue {
  entries: Array<MapResultEntry>
}
interface MapResult extends NormalizeResult {
  value: MapResultValue
}

export type MessageType = 'log' | 'warn' | 'error' | 'info' | 'debug'

export interface Message {
  type: MessageType
  args: Array<any>
}

export function formatMessage(
  type: MessageType,
  args: Array<any | null>
): Message {
  try {
    return {
      type,
      args: formatArgs(args),
    }
  } catch (e) {
    // originalConsole.error(e)
  }
  return {
    type,
    args: [],
  }
}

export function formatArgs(args: Array<any | null>) {
  return args.map((arg) => formatArg(arg))
}

export function formatArg(arg: any | null, depth: number = 0): NormalizeResult {
  if (depth >= 7) {
    return {
      type: 'object',
      value: '[Maximum depth reached]',
    }
  }
  const type = typeof arg
  switch (type) {
    case 'string':
      return formatString(arg as string)
    case 'number':
      return formatNumber(arg as number)
    case 'boolean':
      return formatBoolean(arg as boolean)
    case 'object':
      return formatObject(arg as object, depth)
    case 'undefined':
      return formatUndefined()
    case 'function':
      return formatFunction(arg as Function)
    case 'symbol':
      if (__HARMONY__) {
        return formatUnknown('symbol', arg as unknown)
      } else {
        return formatSymbol(arg as symbol)
      }

    case 'bigint':
      return formatBigInt(arg as unknown)
  }
}

function formatFunction(value: Function): NormalizeResult {
  return {
    type: 'function',
    value: `function ${value.name}() {}`,
  }
}

function formatUndefined(): NormalizeResult {
  return {
    type: 'undefined',
  }
}

function formatBoolean(value: boolean): NormalizeResult {
  return {
    type: 'boolean',
    value: String(value),
  }
}

function formatNumber(value: number): NormalizeResult {
  return {
    type: 'number',
    value: String(value),
  }
}

function formatBigInt(value: unknown): NormalizeResult {
  return {
    type: 'bigint',
    value: String(value),
  }
}

function formatString(value: string): NormalizeResult {
  return {
    type: 'string',
    value,
  }
}

function formatSymbol(value: symbol): NormalizeResult {
  return {
    type: 'symbol',
    value: value.description,
  }
}

function formatUnknown(type: string, value: unknown): NormalizeResult {
  return {
    type,
    value: String(value),
  }
}

function formatObject(value: object, depth: number): NormalizeResult {
  if (value === null) {
    return {
      type: 'null',
    }
  }

  if (!__HARMONY__) {
    if (isComponentPublicInstance(value)) {
      return formatComponentPublicInstance(value, depth)
    }

    if (isComponentInternalInstance(value)) {
      return formatComponentInternalInstance(value, depth)
    }

    if (isUniElement(value)) {
      return formatUniElement(value, depth)
    }

    if (isCSSStyleDeclaration(value)) {
      return formatCSSStyleDeclaration(value, depth)
    }
  }

  if (Array.isArray(value)) {
    return {
      type: 'object',
      subType: 'array',
      value: {
        properties: value.map(
          (v: any | null, i: number): NormalizeResult =>
            formatArrayElement(v, i, depth + 1)
        ),
      },
    } as ArrayResult
  }
  if (value instanceof Set) {
    return {
      type: 'object',
      subType: 'set',
      className: 'Set',
      description: `Set(${value.size})`,
      value: {
        entries: Array.from(value).map(
          (v: any | null): SetResultEntry => formatSetEntry(v, depth + 1)
        ),
      },
    } as SetResult
  }
  if (value instanceof Map) {
    return {
      type: 'object',
      subType: 'map',
      className: 'Map',
      description: `Map(${value.size})`,
      value: {
        entries: Array.from(value.entries()).map(
          (v: Array<any | null>): MapResultEntry => formatMapEntry(v, depth + 1)
        ),
      },
    } as MapResult
  }

  if (value instanceof Promise) {
    return {
      type: 'object',
      subType: 'promise',
      value: {
        properties: [],
      },
    } as ObjectResult
  }
  if (value instanceof RegExp) {
    return {
      type: 'object',
      subType: 'regexp',
      value: String(value),
      className: 'Regexp',
    }
  }
  if (value instanceof Date) {
    return {
      type: 'object',
      subType: 'date',
      value: String(value),
      className: 'Date',
    }
  }
  if (value instanceof Error) {
    return {
      type: 'object',
      subType: 'error',
      value: value.message || String(value),
      className: value.name || 'Error',
    }
  }
  let className: string | undefined = undefined
  if (!__HARMONY__) {
    const constructor = value.constructor
    if (constructor) {
      // @ts-expect-error
      if (constructor.get$UTSMetadata$) {
        // @ts-expect-error
        className = constructor.get$UTSMetadata$().name
      }
    }
  }
  return {
    type: 'object',
    className,
    value: {
      properties: Object.entries(value).map(
        (entry: [string, any | null]): NormalizeResult =>
          formatObjectProperty(entry[0], entry[1], depth + 1)
      ),
    },
  } as ObjectResult
}

function isComponentPublicInstance(
  value: any
): value is ComponentPublicInstance {
  return value.$ && isComponentInternalInstance(value.$)
}

function isComponentInternalInstance(
  value: any
): value is ComponentInternalInstance {
  return value.type && value.uid != null && value.appContext
}

function formatComponentPublicInstance(
  value: ComponentPublicInstance,
  depth: number
) {
  return {
    type: 'object',
    className: 'ComponentPublicInstance',
    value: {
      properties: Object.entries(value.$.type).map(
        ([name, value]): NormalizeResult =>
          formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function formatComponentInternalInstance(
  value: ComponentInternalInstance,
  depth: number
) {
  return {
    type: 'object',
    className: 'ComponentInternalInstance',
    value: {
      properties: Object.entries(value.type).map(
        ([name, value]): NormalizeResult =>
          formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function isUniElement(value: any): value is UniElement {
  return value.style && value.tagName != null && value.nodeName != null
}

function formatUniElement(value: UniElement, depth: number) {
  return {
    type: 'object',
    // 非 x 没有 UniElement 的概念
    // className: 'UniElement',
    value: {
      properties: Object.entries(value)
        .filter(([name]) =>
          [
            'id',
            'tagName',
            'nodeName',
            'dataset',
            'offsetTop',
            'offsetLeft',
            'style',
          ].includes(name)
        )
        .map(
          ([name, value]): NormalizeResult =>
            formatObjectProperty(name, value, depth + 1)
        ),
    },
  }
}

function isCSSStyleDeclaration(
  value: any
): value is CSSStyleDeclaration & { $styles: Record<string, string | null> } {
  return (
    typeof value.getPropertyValue === 'function' &&
    typeof value.setProperty === 'function' &&
    value.$styles
  )
}

function formatCSSStyleDeclaration(
  style: CSSStyleDeclaration & { $styles: Record<string, string | null> },
  depth: number
) {
  return {
    type: 'object',
    value: {
      properties: Object.entries(style.$styles).map(([name, value]) =>
        formatObjectProperty(name, value, depth + 1)
      ),
    },
  }
}

function formatObjectProperty(name: string, value: any | null, depth: number) {
  const result = formatArg(value, depth)
  result.name = name
  return result
}

function formatArrayElement(value: any | null, index: number, depth: number) {
  const result = formatArg(value, depth)
  result.name = `${index}`
  return result
}

function formatSetEntry(value: any | null, depth: number): SetResultEntry {
  return {
    value: formatArg(value, depth),
  }
}

function formatMapEntry(
  value: Array<any | null>,
  depth: number
): MapResultEntry {
  return {
    key: formatArg(value[0], depth),
    value: formatArg(value[1], depth),
  }
}
