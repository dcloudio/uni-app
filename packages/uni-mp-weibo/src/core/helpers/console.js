const unshift = Array.prototype.unshift

function format (args) {
  unshift.call(args, '[system]')
  return args
}

function createLog (method) {
  return function () {
    let printLog = true
    if (method === 'debug' && !__uniConfig.debug) {
      printLog = false
    }
    printLog && global.console[method].apply(global.console, format(arguments))
  }
}

export default {
  log: createLog('log'),
  info: createLog('info'),
  warn: createLog('warn'),
  debug: createLog('debug'),
  error: createLog('error')
}
