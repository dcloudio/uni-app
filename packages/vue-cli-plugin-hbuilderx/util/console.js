const ZERO_WIDTH_CHAR = {
  NOTE: '',
  WARNING: '\u200B',
  ERROR: '\u200C',
  backup0: '\u200D',
  backup1: '\u200E',
  backup2: '\u200F',
  backup3: '\uFEFF'
}

module.exports = {
  warn (msg) {
    msg = ZERO_WIDTH_CHAR.WARNING + msg + ZERO_WIDTH_CHAR.WARNING
    console.warn(msg)
  },
  error (msg) {
    msg = ZERO_WIDTH_CHAR.ERROR + msg + ZERO_WIDTH_CHAR.ERROR
    console.error(msg)
  },
  log (msg) {
    console.log(msg)
  }
}
