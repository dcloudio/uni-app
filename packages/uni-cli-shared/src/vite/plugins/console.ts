import debug from 'debug'
import type { Plugin, ResolvedConfig } from 'vite'
import { type FilterPattern, createFilter } from '@rollup/pluginutils'

import { isJsFile, parseVueRequest } from '../utils'
import { restoreConsoleExpr, rewriteConsoleExpr } from '../../logs/console'
import { withSourcemap } from '../../vite/utils/utils'
import { isRenderjs, isWxs } from '../../filter'

export interface ConsoleOptions {
  method: string
  filename?: (filename: string) => string
  include?: FilterPattern
  exclude?: FilterPattern
}

const debugConsole = debug('uni:console')

export function uniConsolePlugin(options: ConsoleOptions): Plugin {
  const filter = createFilter(options.include, options.exclude)
  let resolvedConfig: ResolvedConfig
  return {
    name: 'uni:console',
    enforce: 'pre',
    configResolved(config) {
      resolvedConfig = config
    },
    transform(code, id) {
      if (isRenderjs(id) || isWxs(id)) {
        return {
          code: restoreConsoleExpr(code),
          map: { mappings: '' },
        }
      }
      if (!filter(id)) return null
      if (!isJsFile(id)) return null
      let { filename } = parseVueRequest(id)
      if (options.filename) {
        filename = options.filename(filename)
      }
      if (!filename) {
        return null
      }
      if (!code.includes('console.')) {
        return null
      }
      debugConsole(id)
      return rewriteConsoleExpr(
        options.method,
        id,
        filename,
        code,
        withSourcemap(resolvedConfig)
      )
    },
  }
}
