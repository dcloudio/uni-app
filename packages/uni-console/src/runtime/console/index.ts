import type { SendFn } from '../utils'
import { type Message, type MessageType, formatMessage } from './format'

const CONSOLE_TYPES = ['log', 'warn', 'error', 'info', 'debug'] as const

let sendConsole: SendFn = null

const messageQueue: Message[] = []

const messageExtra: Record<string, any> = {}

function sendConsoleMessages(messages: Message[]) {
  if (sendConsole == null) {
    messageQueue.push(...messages)
    return
  }
  sendConsole(
    JSON.stringify(
      Object.assign(
        {
          type: 'console',
          data: messages,
        },
        messageExtra
      )
    )
  )
}

export function setSendConsole(value: SendFn, extra: Record<string, any> = {}) {
  sendConsole = value
  Object.assign(messageExtra, extra)
  if (value != null && messageQueue.length > 0) {
    const messages = messageQueue.slice()
    messageQueue.length = 0
    sendConsoleMessages(messages)
  }
}

export const originalConsole = /*@__PURE__*/ CONSOLE_TYPES.reduce(
  (methods, type) => {
    methods[type] = console[type].bind(console)
    return methods
  },
  {} as Record<MessageType, typeof console.log>
)

const atFileRegex = /^\s*at\s+[\w/./-]+:\d+$/

export function rewriteConsole() {
  function wrapConsole(type: MessageType) {
    return function (...args: any[]) {
      const originalArgs = [...args]
      if (originalArgs.length) {
        const maybeAtFile = originalArgs[originalArgs.length - 1]
        // 移除最后的 at pages/index/index.uvue:6
        if (typeof maybeAtFile === 'string' && atFileRegex.test(maybeAtFile)) {
          originalArgs.pop()
        }
      }
      if (process.env.UNI_CONSOLE_KEEP_ORIGINAL) {
        originalConsole[type](...originalArgs)
      }
      sendConsoleMessages([formatMessage(type, args)])
    }
  }

  // 百度小程序不允许赋值，所以需要判断是否可写
  if (isConsoleWritable()) {
    CONSOLE_TYPES.forEach((type) => {
      console[type] = wrapConsole(type)
    })
    return function restoreConsole() {
      CONSOLE_TYPES.forEach((type) => {
        console[type] = originalConsole[type]
      })
    }
  } else {
    if (!process.env.UNI_CONSOLE_WEBVIEW) {
      if (typeof uni !== 'undefined' && uni.__f__) {
        const oldLog = uni.__f__
        if (oldLog) {
          // 重写 uni.__f__ 方法，这样的话，仅能打印开发者代码里的日志，其他没有被重写为__f__的日志将无法打印（比如uni-app框架、小程序框架等）
          uni.__f__ = function (...args: any[]) {
            const [type, filename, ...rest] = args
            // 原始日志移除 filename
            oldLog(type, '', ...rest)
            sendConsoleMessages([formatMessage(type, [...rest, filename])])
          }
          return function restoreConsole() {
            uni.__f__ = oldLog
          }
        }
      }
    }
  }
  return function restoreConsole() {}
}

function isConsoleWritable() {
  const value = console.log
  const sym = Symbol()
  try {
    // @ts-expect-error
    console.log = sym
  } catch (ex) {
    return false
  }
  // @ts-expect-error
  const isWritable = console.log === sym
  console.log = value
  return isWritable
}
