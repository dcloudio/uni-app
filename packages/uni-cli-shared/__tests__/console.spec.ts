import { rewriteConsoleExpr } from '../src/logs/console'
const filename = 'foo.vue'
const METHOD = '__f__'
describe('console', () => {
  test('console.log', () => {
    expect(
      rewriteConsoleExpr(
        METHOD,
        filename,
        filename,
        `const a = 1;console.log(a);`
      ).code
    ).toMatchSnapshot()
  })
  test('console.log multiline', () => {
    expect(
      rewriteConsoleExpr(
        METHOD,
        filename,
        filename,
        `const a = 1;

console.log(a);
const b = 2
console.log(a,b);
console.log(a,b,c);
`
      ).code
    ).toMatchSnapshot()
  })
  test('console.info', () => {
    expect(
      rewriteConsoleExpr(METHOD, filename, filename, `console.info(a,b,c);`)
        .code
    ).toMatchSnapshot()
  })
  test('console.debug', () => {
    expect(
      rewriteConsoleExpr(METHOD, filename, filename, `console.info(a,b,c);`)
        .code
    ).toMatchSnapshot()
  })
  test('console.warn', () => {
    expect(
      rewriteConsoleExpr(METHOD, filename, filename, `console.info(a,b,c);`)
        .code
    ).toMatchSnapshot()
  })
  test('console.error', () => {
    expect(
      rewriteConsoleExpr(METHOD, filename, filename, `console.info(a,b,c);`)
        .code
    ).toMatchSnapshot()
  })
})
