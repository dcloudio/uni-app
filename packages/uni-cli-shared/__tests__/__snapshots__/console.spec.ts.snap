// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`console appendConsoleExpr 1`] = `
"// 基础测试用例
console.log('简单字符串', " at test.ts:2")
console.info('双引号字符串', " at test.ts:3")
console.warn(\`模板字符串\`, " at test.ts:4")
console.error('带分号的', " at test.ts:5")
console.debug('不带分号的', " at test.ts:6")

// 多参数测试
console.log('参数1', '参数2', '参数3', " at test.ts:9")
console.info(1, true, null, undefined, " at test.ts:10")
console.warn('混合类型', 123, true, null, " at test.ts:11")

// 对象和数组测试
console.log({ name: 'test', age: 18 }, " at test.ts:14")
console.log(['a', 'b', 'c'], " at test.ts:15")
console.log({ a: 1 }, ['x', 'y'], { b: 2 }, " at test.ts:16")

// 表达式测试
console.log(1 + 2, " at test.ts:19")
console.log(foo(), " at test.ts:20")
console.log(a ? b : c, " at test.ts:21")
console.log(obj.method(), " at test.ts:22")

// 模板字符串测试
console.log(\`Hello \${name}\`, " at test.ts:25")
console.log(\`多行
模板字符串
测试\`, " at test.ts:26")
console.log(\`\${obj.method()} 和 \${arr[0]}\`, " at test.ts:29")

// 多行格式测试
console.log('多行参数1', '多行参数2', '多行参数3', " at test.ts:32")

console.log({
  name: 'test',
  age: 18,
  nested: {
    a: 1,
    b: 2,
  },
}, " at test.ts:34")

// 嵌套括号测试
console.log(getData('test'), " at test.ts:44")
console.log(outer(inner()), " at test.ts:45")
console.log((a + b) * c, " at test.ts:46")

// 函数调用测试
console.log(foo.bar(), " at test.ts:49")
console.log(new Date(), " at test.ts:50")
console.log(Array.from(set), " at test.ts:51")

// 复杂混合测试
console.log(\`用户 \${user.name}\`,
  {
    id: user.id,
    time: new Date(),
    data: getData(user.id),
  },
  process.env.NODE_ENV === 'development' && '开发环境', " at test.ts:54")

// 注释和特殊字符测试
console.log("包含'单引号'", " at test.ts:65")
console.log('包含"双引号"', " at test.ts:66")
console.log(\`包含\\\`反引号\\\`\`, " at test.ts:67")
console.log('包含/正斜杠/', " at test.ts:68")
console.log('包含\\\\反斜杠\\\\', " at test.ts:69")

// 正则表达式测试
console.log(/test/g, " at test.ts:72")
console.log(new RegExp('test', 'g'), " at test.ts:73")

// 特殊语法测试
console.log(...args, " at test.ts:76")
console.log?.('optional chaining')

// 带注释的测试
console.log('test', // 行尾注释
  /* 块注释 */ 'test2',
  'test3' /* 行内块注释 */, " at test.ts:80")

// 空参数测试
console.log(, " at test.ts:87")
console.log('', " at test.ts:88")
console.log(, " at test.ts:89")

// 连续调用测试
console.log('test1', " at test.ts:92")
console.log('test2', " at test.ts:93")
console.log('test1', " at test.ts:94")
console.log('test2', " at test.ts:95")

// 条件语句中的测试
if (true) console.log('条件语句', " at test.ts:98")
if (true) {
  console.log('条件语句块', " at test.ts:100")
}

// 箭头函数中的测试
const fn = () => console.log('箭头函数', " at test.ts:104")
const fn2 = () => {
  console.log('箭头函数块', " at test.ts:106")
}

console.log(\`console.log('test')\`, " at test.ts:109")
"
`;

exports[`console console.debug 1`] = `"__f__('info','at foo.vue:1',a,b,c);"`;

exports[`console console.error 1`] = `"__f__('info','at foo.vue:1',a,b,c);"`;

exports[`console console.info 1`] = `"__f__('info','at foo.vue:1',a,b,c);"`;

exports[`console console.log 1`] = `"const a = 1;__f__('log','at foo.vue:1',a, JSON.stringify(1));"`;

exports[`console console.log multiline 1`] = `
"const a = 1;

__f__('log','at foo.vue:3',a);
const b = 2
__f__('log','at foo.vue:5',a,b);
__f__('log','at foo.vue:6',a,b,c);
"
`;

exports[`console console.warn 1`] = `"__f__('info','at foo.vue:1',a,b,c);"`;
